62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,801 @@
+            segment.t_start = 0.0;
+            segment.t_end = 1.0;
+            segments.push_back(segment);
+        }
+
+        std::sort(segments.begin(), segments.end(), [&](const Segment& a, const Segment& b) { return a.originalBezier->p[0][major] > b.originalBezier->p[0][major]; });
+		for (Segment& segment : segments)
+		{
+			starts.push(segment);
+			//std::cout << "Starts: " << segment.originalBezier->p[0][major]  << "\n";
+		}
+
+        std::sort(segments.begin(), segments.end(), [&](const Segment& a, const Segment& b) { return a.originalBezier->p[2][major] > b.originalBezier->p[2][major]; });
+        for (Segment& segment : segments)
+		{
+			ends.push(segment.originalBezier->p[2][major]);
+			//std::cout << "Ends: " << segment.originalBezier->p[2][major] << "\n";
+		}
+        maxMajor = segments.front().originalBezier->p[2][major];
+    }
+
+    // Sweep line algorithm
+    std::vector<Segment> activeCandidates; // Set of active candidates for neighbor search in sweep line
+
+	// if we weren't spawning quads, we could just have unsorted `vector<Bezier*>`
+	auto candidateComparator = [&](const Segment& lhs, const Segment& rhs)
+	{
+		// btw you probably want the beziers in Quadratic At^2+B+C form, not control points
+		double _lhs = lhs.originalBezier->evaluateBezier(lhs.t_start)[minor];
+		double _rhs = rhs.originalBezier->evaluateBezier(rhs.t_start)[minor];
+
+		double lenLhs = glm::distance(lhs.originalBezier->p[0], lhs.originalBezier->p[2]);
+		double lenRhs = glm::distance(rhs.originalBezier->p[0], rhs.originalBezier->p[2]);
+		auto minLen = std::min(lenLhs, lenRhs);
+
+		//printf("Comparing bezier lhs = (%f, %f)..(%f, %f) rhs = (%f, %f)..(%f, %f) || minor at t_start (%f, %f) = %f < %f || scale of the curves: lhs = %f rhs = %f minLen = %f || abs(_lhs - _rhs) / minLen = %f\n",
+		//	lhs.originalBezier->p[0].x, lhs.originalBezier->p[0].y,
+		//	lhs.originalBezier->p[2].x, lhs.originalBezier->p[2].y,
+		//	rhs.originalBezier->p[0].x, rhs.originalBezier->p[0].y,
+		//	rhs.originalBezier->p[2].x, rhs.originalBezier->p[2].y,
+		//	lhs.t_start, rhs.t_start,
+		//	_lhs, _rhs,
+		//	len(lhs.originalBezier->p[2] - lhs.originalBezier->p[0]),
+		//	len(rhs.originalBezier->p[2] - rhs.originalBezier->p[0]),
+		//	minLen,
+		//	abs(_lhs - _rhs) / minLen
+		//);
+		// Threshhold here for intersection points, where the minor values for the curves are
+		// very close but could be smaller, causing the curves to be in the wrong order
+		// TODO: figure out if this is the best ay to do this
+		if (abs(_lhs - _rhs) < 1e-3 * minLen)
+		{
+			// this is how you want to order the derivatives dmin/dmaj=-INF dmin/dmaj = 0 dmin/dmaj=INF
+			// also leverage the guarantee that `dmaj>=0` to ger numerically stable compare
+			float64_t2 lTan = lhs.originalBezier->tangent(lhs.t_start);
+			float64_t2 rTan = rhs.originalBezier->tangent(rhs.t_start);
+			_lhs = lTan[minor] * rTan[major];
+			_rhs = rTan[minor] * lTan[major];
```

Comment:
assert here than `rTan[major]` and `lTan[major]` are >= because negative values mess with the comparison operator when we do mul and they should be positive because of major monotonicity